File: coding-test-1/practice-problems/solutions/01-factorial.rb
# Write a method that takes an integer `n` in; it should return
# n*(n-1)*(n-2)*...*2*1. Assume n >= 0.
#
# As a special case, `factorial(0) == 1`.
#
# Difficulty: easy.

def factorial(n)
  if n < 0
    return nil
  end

  result = 1
  while n > 0
    result = result * n

    n -= 1
  end

  return result
end

puts("factorial(0) == 1: #{factorial(0) == 1}")
puts("factorial(1) == 1: #{factorial(1) == 1}")
puts("factorial(2) == 2: #{factorial(2) == 2}")
puts("factorial(3) == 6: #{factorial(3) == 6}")
puts("factorial(4) == 24: #{factorial(4) == 24}")

File: coding-test-1/practice-problems/solutions-python/01-factorial.py
# Write a method that takes an integer `n` in; it should return
# n*(n-1)*(n-2)*...*2*1. Assume n >= 0.
#
# As a special case, `factorial(0) == 1`.
#
# Difficulty: easy.
import unittest


def factorial(n):

    if n < 0:
        return None

    result = 1
    while n > 0:
        result *= n
        n -= 1

    return result

# These are tests to check that your code is working. After writing
# your solution, they should all run.


class TestFactorial(unittest.TestCase):

    def test_zero_case(self):
        self.assertEqual(factorial(0), 1)

    def test_simple_case(self):
        self.assertEqual(factorial(1), 1)

    def test_simple_case_2(self):
        self.assertEqual(factorial(2), 2)

    def test_simple_case_3(self):
        self.assertEqual(factorial(3), 6)

    def test_simple_case_4(self):
        self.assertEqual(factorial(4), 24)

if __name__ == '__main__':
    unittest.main()